### PR TITLE
Update the `qbraid.runtime` with unified experiment results

### LINKED ISSUE #743: Create unified `qbraid.runtime.Result` class
Currently we have distinct result classes for each runtime provider. However, the function of the classes if effectively the same, in that they each implement a method `get_counts` that gathers the histogram data, and then methods within the parent `GateModelJobResult` class are used to format and return the measurement counts.

Instead, the functionality of the `get_counts` method should be moved to an internal method in each of the providers Job classes. Then we should gather the counts by calling the method from within the `QuantumJob.result` method, and return a unified `qbraid.runtime.Result` object. The counts can then be formatted as needed during initialization or when the user first tries to retrieve the counts. This will also allow us to make the result class a bit more rigid, as right now it accepts a single unspecified object. So instead, it will now allow us to define distinct arguments and include more data about the run e.g. job_id, shots, etc. This will hopefully make the job $\rightarrow$ result flow easier to understand for 3rd parties implementing these classes.

The inspiration for this "unified" result class comes from the [qiskit.result.Result](https://docs.quantum.ibm.com/api/qiskit/qiskit.result.Result) class, which I encourage you to use as reference.

See also:
- https://github.com/Qiskit/qiskit/blob/stable/1.2/qiskit/result/result.py
- https://github.com/Qiskit/qiskit/blob/stable/1.2/qiskit/result/models.py
- https://github.com/Qiskit/qiskit/blob/stable/1.2/qiskit/result/counts.py
- https://github.com/Alice-Bob-SW/qiskit-alice-bob-provider/blob/c5847cbf31880f214b8467432207eb665c832778/qiskit_alice_bob_provider/remote/job.py#L235-L259
- https://github.com/microsoft/azure-quantum-python/blob/b43e101793156f9dd50a0c57372619bb5e2a49ec/azure-quantum/azure/quantum/qiskit/job.py#L82-L102

Similar to qiskit, our `qbraid.runtime.Result` class could take in universal metadata like job_id, device_id, success, etc. as well as a single (or list of) `ExperimentalResult` objects. Sort of similar to what we already do for the [QbraidJobResult](https://github.com/qBraid/qBraid/blob/f0edee0f468b354d18b002cfae87f35528e67016/qbraid/runtime/native/result.py#L46-L49) class. However, unlike the `qiskit.result.ExperimentResult`, and unlike the current `qbraid.runtime.ExperimentalResult`, this new `ExperimentalResult` could be a shared type to 2 or more sub-classes that account for any differences in device modality, and how the results may need to be processed. For example, Amazon Braket defines 3 result types, and the [AwsQuantumTask.result](https://github.com/amazon-braket/amazon-braket-sdk-python/blob/8f4e88f3a6389306fdfb6f64866b0eea9ebe8db4/src/braket/aws/aws_quantum_task.py#L396-L426) method returns the `Union` of these types. We were forced to mirror this structure with our own [BraketAhsJobResult](https://github.com/qBraid/qBraid/blob/f0edee0f468b354d18b002cfae87f35528e67016/qbraid/runtime/braket/result.py#L64) and [BraketGateModelJobResult](https://github.com/qBraid/qBraid/blob/f0edee0f468b354d18b002cfae87f35528e67016/qbraid/runtime/braket/result.py#L40). If we were to continue with this current structure, we would not only end up with a separate result class for each provider, but potentially multiple result classes for each provider.

Instead, we would like to create a results structure that is the best of both worlds between the unified, but rigid, qiskit approach, and the fragmented, but flexible Amazon Braket approach. This could look something like the following:

```python
class ExperimentalResult:
    ...

class GateModelExperimentalResult(ExperimentalResult):
    ...

class AhsExperimentalResult(ExperimentalResult):
    ...

class Result:

    def __init__(job_id, device_id, success, list[ExperimentalResult], ...):
        ...

```

Alternatively, we could maintain a single `ExperimentalResult` class, and have a separate `ResultBuilder` class that can be implemented providers that require distinct result handling for different devices, similar to what we implemented for the [qbraid.runtime.azure.AzureResultBuilder](https://github.com/qBraid/qBraid/blob/f1b794b52df7c9ead34bb0cb873c29f3cf36c5c0/qbraid/runtime/azure/result_builder.py#L50), which can be used to streamline the processing of data within the job class as follows:

https://github.com/qBraid/qBraid/blob/f1b794b52df7c9ead34bb0cb873c29f3cf36c5c0/qbraid/runtime/azure/job.py#L76-L87

It is safe to say that there are many different approaches to how this can be done. However, the main takeaway from this issue is that we would like every provider's `QuantumJob.result` method to return a unified `qbraid.runtime.Result` object, but the way that you choose to structure that class is up to you. Take inspiration from what other providers have done well and what can improved, and keep the end-user in mind, as well as 3rd party developers who may at some point be using the abstractions in the own pipelines in the future.

### PR DESCRIPTION
**Pull request by TheGupta2012**
## Summary of changes
Fixes #743 

- The `qbraid.runtime.result` defines `QuantumJobResult` and subsequent `GateModelJobResult` classes. These are inherited in different providers to define provider specific results
- As mentioned in #743 , this leads to new result classes pertaining to each provider and possibly multiple such definitions
- A new approach to define these results is to build them based on the "quantum experiment types" and let the providers deal with how they aim to build these result objects.
- Currently we define two types of results and their respective handling functions- 
    - `GATE_MODEL` and `_build_runtime_gate_model_results`
    - `AHS` and `_build_runtime_ahs_results`

- Tasks- 
     - [x] `runtime.native`
     - [x] `runtime.braket`
     - [x] `runtime.azure`
     - [x] `runtime.ionq`
     - [x] `runtime.oqc`
     - [x] `runtime.qiskit`
     - [x] remove old functionality
     - [ ] add code coverage

### CODE REVIEW

File: qbraid/runtime/native/job.py [hunk 1]
```diff
@@ -85,11 +84,29 @@ def cancel(self) -> None:
...
+        Args:
+            job_data (dict): The job data dictionary.
+        """
+        measurement_counts = job_data.get("measurementCounts", {})
+        time_stamps: dict[str, Any] = job_data.get("timeStamps", {})
+        execution_duration: int = time_stamps.get("executionDuration", -1)
+        return [
+            ExperimentalResult(
+                state_counts=measurement_counts,
+                measurements=ResultFormatter.counts_to_measurements(measurement_counts),
```
Comment (ryanhill1) [inline qbraid/runtime/native/job.py:99]: he explicit measurements are only needed if we don't have histogram counts data. If we only have the measurements array and no counts data, then we use the measurements to create the histogram counts. After we have the counts data, the full measurements array is no longer needed. If provided, we can still store it. But in my mind it doesn't make a lot of sense to go out of our way to "recreate" them from the counts if not provided

**Review by ryanhill1 — CHANGES_REQUESTED** [review]

File: qbraid/runtime/result.py [hunk 2]
```diff
@@ -139,4 +98,233 @@ def measurement_counts(
...
+        if len(batch_counts) == 1:
+            return batch_counts[0]
+
+        return ResultFormatter.normalize_batch_bit_lengths(batch_counts)
+
+    def measurement_probabilities(self, **kwargs) -> dict[str, float]:
+        """Calculate and return the probabilities of each measurement result."""
+        counts = self.measurement_counts(**kwargs)
+        probabilities = ResultFormatter.counts_to_probabilities(counts)
+        return probabilities
```
Comment (ryanhill1) [inline qbraid/runtime/result.py:330]: There should be no further structural manipulation of data within the `RuntimeJobResult` class. Instead, job classes should utilize the `ResultFormatter` or similar methods to process the raw results data, such as generating the probability dictionary. These processed results can then be passed and accessed directly through the `RuntimeJobResult`.

Alternatively, you could implement a `@classmethod`, like `RuntimeJobResult.from_dict`, which accepts a dictionary containing raw measurement counts or probabilities, processes them into qBraid-compatible formats, and then creates a new `RuntimeJobResult` with this structured data. For reference, you can look at [qiskit.result.Result.from_dict](https://github.com/Qiskit/qiskit/blob/stable/1.2/qiskit/result/result.py#L112-L131).

Non-structural manipulations, such as formatting or rounding (like you've used in the `measurement_counts` method), *are* acceptable within the `RuntimeJobResult` class. For example, the [qiskit.result.postprocess](https://github.com/Qiskit/qiskit/blob/stable/1.2/qiskit/result/postprocess.py) functions in `qiskit` handle such "post-processing" tasks—these involve aesthetic changes like formatting strings or adjusting precision. 

In summary:

- **Post-processing functions** that modify the formatting of data but return the same data type as the input can remain within the `RuntimeJobResult` class. *These functions should also cache their results to avoid repeated processing.*
- **Data transformations** that create new data types or transform input into different output types (e.g., generating probabilities from measurement counts) should occur before the initialization of `RuntimeJobResult`.

**Review by ryanhill1 — CHANGES_REQUESTED** [review]

File: tests/runtime/test_ionq_runtime.py [hunk 3]
```diff
@@ -92,7 +93,7 @@
     "id": "c86a043a-6aea-47cf-b3a6-70ab1e538cab",
     "submitted_by": "e1254d93988e405e80d7842a",
     "status": "completed",
-    "target": "simulator",
+    "tar    get": "simulator",
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply tests/runtime/test_ionq_runtime.py:96]: will fix, must be a typo

File: qbraid/runtime/result.py [hunk 4]
```diff
@@ -139,4 +98,233 @@ def measurement_counts(
...
+
+# Experimental types should define the nature of classes, not providers
+@dataclass
+class ExperimentalResult:
+    """Class to represent the results of a quantum circuit experiment."""
+
+    state_counts: dict
+    measurements: Union[np.ndarray, Any]  # if gate_model_type it would be measurement counts
+    # if AHS it would be the energy levels
+    result_type: ExperimentType = ExperimentType.GATE_MODEL
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply qbraid/runtime/result.py:112]: Okay, got it I'll remove the default. 
I reckoned that a majority of the providers used GateModel and that's why updated that as default value.

File: qbraid/runtime/result.py [hunk 5]
```diff
@@ -139,4 +98,233 @@ def measurement_counts(
             for counts in get_counts
         ]
 
-        return normalize_batch_bit_lengths(batch_counts)
+        return ResultFormatter.normalize_batch_bit_lengths(batch_counts)
+
+
+# Experimental types should define the nature of classes, not providers
+@dataclass
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply qbraid/runtime/result.py:105]: Okay, I'll see if it is growing a lot will change to a relevant class
